cmds/db-manager/main.go

```diff
@@ -0,0 +1,219 @@
+// Bootstrap script for Database deployment and migration
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"math"
+	"os"
+	"os/signal"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/golang-migrate/migrate/v4"
+	_ "github.com/golang-migrate/migrate/v4/database/cockroachdb"
+	_ "github.com/golang-migrate/migrate/v4/source/file"
+	dssCockroach "github.com/interuss/dss/pkg/cockroach"
+	"go.uber.org/zap"
+)
+
+// MyMigrate is an alias for exstending migrate.Migrate
+type MyMigrate struct {
+	*migrate.Migrate
+}
+
+// Direction is an alias for int indicating the direction and steps of migration
+type Direction int
+
+func (d Direction) String() string {
+	if d > 0 {
+		return "Up"
+	} else if d < 0 {
+		return "Down"
+	}
+	return "No Change"
+}
+
+var (
+	path      = flag.String("migration_dir", "", "path to db migration files directory")
+	dbVersion = flag.String("db_version", "", "the db version to migrate to (ex: v1.0.0)")
+	step      = flag.Int("migration_step", 0, "the db migration step to go to")
+
+	cockroachParams = struct {
+		host            *string
+		port            *int
+		sslMode         *string
+		sslDir          *string
+		user            *string
+		applicationName *string
+	}{
+		host:    flag.String("cockroach_host", "0.0.0.0", "cockroach host to connect to"),
+		port:    flag.Int("cockroach_port", 26257, "cockroach port to connect to"),
+		sslMode: flag.String("cockroach_ssl_mode", "disable", "cockroach sslmode"),
+		user:    flag.String("cockroach_user", "root", "cockroach user to authenticate as"),
+		sslDir:  flag.String("cockroach_ssl_dir", "", "directory to ssl certificates. Must contain files: ca.crt, client.<user>.crt, client.<user>.key"),
+	}
+)
+
+func main() {
+	flag.Parse()
+	if *path == "" {
+		log.Panic("Must specify migration_dir path")
+	}
+	if (*dbVersion == "" && *step == 0) || (*dbVersion != "" && *step != 0) {
+		log.Panic("Must specify a db_version or migration_step to goto")
+	}
+	uriParams := map[string]string{
+		"host":             *cockroachParams.host,
+		"port":             strconv.Itoa(*cockroachParams.port),
+		"user":             *cockroachParams.user,
+		"ssl_mode":         *cockroachParams.sslMode,
+		"ssl_dir":          *cockroachParams.sslDir,
+		"application_name": "SchemaManager",
+	}
+	postgresURI, err := dssCockroach.BuildURI(uriParams)
+	if err != nil {
+		log.Panic("Failed to build URI", zap.Error(err))
+	}
+	crdbURI := strings.Replace(postgresURI, "postgresql", "cockroachdb", 1)
+	*path = fmt.Sprintf("file://%v", *path)
+	migrater, migraterErr := migrate.New(*path, crdbURI)
+	myMigrater := &MyMigrate{migrater}
```

Comment:
there is a check for migrateErr != nil a few lines below. Creating this "MyMigrate" object makes the call more uniform with the deferred function

Reply:
Yep; it seems like this could be much simpler with:
```go
migrater, err := migrate.New(*path, crdbURI)
if err != nil {
  log.Panic(err)
}
myMigrater := &MyMigrate{migrater}
defer func() {
	if _, err := myMigrater.Close(); err != nil {
		log.Println(err)
	}
}()
```
(then no consideration of migrateErr anywhere else)

As written currently, it seems like an error on line 84 will result in an invalid migrater, yet line 109 and beyond will still run resulting in undefined behavior, so basically I'm asking here whether it is valid to do anything else in the routine if migrateErr is not nil (therefore implying that migrater is not valid).